[wip] Statically built Nix

matthewbauer opened the pull request:
###### Motivation for this change

This makes it possible to build Nix statically through pkgsStatic.nix. Needs a couple of patches to nix but otherwise works smoothly.

If you'd like to try it, you can just check out this branch and run:

```
nix-build -A pkgsStatic.nix
```

and wait a little bit. More targets are available here:

https://github.com/matthewbauer/static-nix

Alternatively, you can use my prebuilt binary at https://matthewbauer.us/nix. It is built for x86_64-linux and can be used on any system like this:

```
$ curl https://matthewbauer.us/nix | sh -s ...
```

For instance, on a linux system without nix installed, you can do this:

```
$ curl https://matthewbauer.us/nix | sh -s run --store $HOME/.cache/nix/store -f channel:nixpkgs-unstable hello -c hello
Hello World!
```

There are many uses cases where nix is unavailable and the above command can be extremely useful. It's a little slow still, but can be sped up by reusing the same ```--store``` value (such as $HOME/.cache/nix/store). Here is the time:

```
$ time (curl https://matthewbauer.us/nix | sh -s run --store $HOME/.cache/nix/store -f channel:nixpkgs-unstable hello -c hello)
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 5596k  100 5596k    0     0  2807k      0  0:00:01  0:00:01 --:--:-- 2807k
[2 copied (24.0 MiB), 35.5 MiB DL]
Hello, world!
( curl https://matthewbauer.us/nix | sh -s run --store $(mktemp -d) -f  hello)  3.39s user 3.52s system 70% cpu 9.757 total
```

matthewbauer committed:
stdenvAdapters: don’t separate debug info with static binaries

separateDebugInfo doesn’t seem to work with makeStaticBinaries.

matthewbauer committed:
busybox: fix musl builds

You shouldn’t need to add anything with musl builds. The libc will
configure these values for you.

matthewbauer committed:
krb5: remove static handling

This never worked due to krb5 using dynamic linking for plugin
functionality. No reason in falsely advertisting this works. See this
thread for more discussion:

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=439039

matthewbauer committed:
boehm-gc: support static + musl

matthewbauer committed:
static: support some more libraries

adds support for boost, brotli, and libcurl.

matthewbauer committed:
brotli: add staticOnly option

matthewbauer committed:
nix: fix static building

aws libs fail to link statically

nix static add -latomic

ryantm commented:
I tried following some of the examples in your [blog post](https://matthewbauer.us/blog/static-nix.html) on my NixOS machine, and I get this error message:

```
[ryantm@laptop3:~]$ $ curl https://matthewbauer.us/nix | sh -s run --store $HOME/.cache/nix/store -f channel:nixpkgs-unstable hello -c hello
$: command not found
[ryantm@laptop3:~]$ curl https://matthewbauer.us/nix | sh -s run --store $HOME/.cache/nix/store -f channel:nixpkgs-unstable hello -c hello
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 5596k  100 5596k    0     0  1408k      0  0:00:03  0:00:03 --:--:-- 1408k
error: getting status of '/nix/store/p492jkrs1q0nddjqghw5d5s5w7kgifda-nix-2.2-x86_64-unknown-linux-musl': No such file or directory
[ryantm@laptop3:~]$ nix=$(mktemp); \
>   curl https://matthewbauer.us/nix > $nix && \
>   chmod +x $nix && \
>   $nix run --store $HOME/.cache/nix/store -f channel:nixpkgs-unstable \
>   ranger -c ranger
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 5596k  100 5596k    0     0  5264k      0  0:00:01  0:00:01 --:--:-- 5269k
error: getting status of '/nix/store/p492jkrs1q0nddjqghw5d5s5w7kgifda-nix-2.2-x86_64-unknown-linux-musl': No such file or directory
```

jb55 commented:
I'm getting the same error as @ryantm

matthewbauer commented:
Thanks for the heads up. I had forgotten about those extra files needed. The updated binary at ```https://matthewbauer.us/nix``` should have a working copy.

For future reference, this is how it was built:

```
nix-build -A pkgsStatic.nix
mkdir -p nix
cp ./result/bin/nix nix/nix
cp -r ./result/share nix/share
cd nix
chmod 755 chmod 755 share/ share/nix/ share/nix/corepkgs/
rm -rf share/nix/sandbox
chmod 644 nix share/nix/corepkgs/*
tar cfz nix.tar.gz nix share/
arx tmpx ./nix.tar.gz // './nix "$@"' > ./nix.sh
cp ./nix.sh ~/matthewbauer.github.io/nix
```

ryantm commented:
Works now for me!

bhipple commented on pkgs/top-level/static.nix:
```diff
@@ -151,4 +155,19 @@ in {
+
+  curl = super.curl.override {
+    gssSupport = false;
+  };
```
I would imagine your target audience for this is users who are running and deploying on rootless "enterprise" environments such a RHEL or CentOS.  It's worth noting that in many of these setups, kerberos is used extensively for all HTTP and git authentication, so if you could get it working in this static variant it'd probably be worthwhile.

matthewbauer replied:
Yeah this is an issue. libkrb5 doesn't have good way to link statically, however:

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=439039

kerberos heimdal might support it though.

bhipple commented on pkgs/top-level/static.nix:
```diff
@@ -151,4 +155,19 @@ in {
+  };
+
+  nix = super.nix.overrideAttrs (_: {
+    NIX_LDFLAGS = "-lssl -lbrotlicommon -lssh2 -lz -lnghttp2 -lcrypto";
```
I think the right way to do this would be patching the pkg-config files to include these libraries in the `Requires.private` section, as is done in:

```
$ nix-build -A brotli.dev && cat result-dev/lib/pkgconfig/libbrotlidec.pc
prefix=/nix/store/7i965pyrkqph3gwpfzswz35c6a1j1v5h-brotli-1.0.7
exec_prefix=/nix/store/7i965pyrkqph3gwpfzswz35c6a1j1v5h-brotli-1.0.7
libdir=/nix/store/iahrxydkznvf6z0q0hp0cp9nhrq1pkys-brotli-1.0.7-lib/lib
includedir=/nix/store/1qcykjlmxvvzgvvnyzsz7yii794m5rjv-brotli-1.0.7-dev/include

Name: libbrotlidec
URL: https://github.com/google/brotli
Description: Brotli decoder library
Version: 1.0.7
Libs: -L${libdir} -lbrotlidec
Requires.private: libbrotlicommon >= 1.0.2
Cflags: -I${includedir}
```
I haven't looked at this specific case too closely, but in general invoking `pkgconfig --static <args>` will also include the `Requires.private` sections for transitive link dependencies that are not needed in the dynamic variant.

matthewbauer replied:
I think the issue was that all of these need to be propagated in static. static linking requires knowing where all of the needed libraries are.

edolstra commented:
In principle this looks fine to me, but it clutters Nixpkgs with numerous patches and special-case attributes that are going to be a pain in the ass to maintain. Who is going to be responsible for maintaining those?

ryantm commented:
> Who is going to be responsible for maintaining those?

One possible part of the solution would be for @matthewbauer to add himself as a maintainer of these packages.

dtzWill commented on pkgs/development/libraries/boehm-gc/default.nix:
```diff
@@ -24,14 +24,17 @@ stdenv.mkDerivation rec {
+    name = "boehm-gc-7.6.0-sys_select.patch";
+    url = "https://gitweb.gentoo.org/proj/musl.git/plain/dev-libs/boehm-gc/files/boehm-gc-7.6.0-sys_select.patch?id=85b6a600996bdd71162b357e9ba93d8559342432";
+    sha256 = "1gydwlklvci30f5dpp5ccw2p2qpph5y41r55wx9idamjlq66fbb3";
+  }) ] ++
```
Is this still needed? It was explicitly removed in a recent-ish PR with the idea that this didn't hurt but didn't help, IIRC.  Might be good to check.

dtzWill commented on pkgs/stdenv/adapters.nix:
```diff
@@ -44,6 +44,7 @@ rec {
       then throw "Cannot build fully static binaries on Darwin/macOS"
       else stdenv'.mkDerivation (args // {
         NIX_CFLAGS_LINK = toString (args.NIX_CFLAGS_LINK or "") + " -static";
+        separateDebugInfo = false;
```
This doesn't seem right, why must this be disabled for static adapter?

matthewbauer replied:
Yeah maybe not... I was getting an issue that I thought was due to this but it might have been unrelated.

matthewbauer commented:
Yeah i can do that. Lots of these are because of ad hoc configure scripts (no autotools, cmake, or meson). Other times static libs are not being installed correctly into prefix. All of these can probably be corrected with upstream patches though.

orivej commented on pkgs/development/libraries/aws-c-common/default.nix:
```diff
@@ -13,6 +13,10 @@ stdenv.mkDerivation rec {
   nativeBuildInputs = [ cmake ];
 
+  postPatch = stdenv.lib.optionalString stdenv.hostPlatform.isMusl ''
+    sed -i "24ireturn()" cmake/AwsTestHarness.cmake
```
This is fragile. In general, a better approach is to generate a patch with `diff -U1`.

This breaks non-musl configure with:
```
CMake Error at tests/CMakeLists.txt:260 (target_compile_options):
  Cannot specify compile options for target "aws-c-common-tests" which is not
  built by this project.
```
Why does it work with musl?

You could try `-DBUILD_TESTING=OFF` instead.

matthewbauer replied:
Yeah it’s super broken. BUILD_TESTING doesn’t work at least

orivej replied:
I see, `BUILD_TESTING` is not actually implemented. You could patch out `add_subdirectory(tests)` in the main `CMakeLists.txt` then.

illegalprime commented:
This looks really cool, it would help people (me) run nix on ARM with limited memory / disk space (by downloading a static nix and loading some external disk space and setting the nix store inside there).
Is this the right derivation to build?
```
nix build -f . pkgsStatic.pkgsCross.aarch64-multiplatform.nix
```

matthewbauer commented:
I don't think that will work - it will build a static toolchain and then cross compile with it. Ideally, you would pass these args to Nixpkgs:

```
nix build -f. \
  --arg crossSystem '{ config = "aarch64-unknown-linux-musl"; }' \
  --arg crossOverlays '[ (import ./pkgs/top-level/static.nix) ]'
```

But I don't think crossOverlays is exposed yet.

matthewbauer committed:
top-level/impure.nix: expose crossOverlays

illegalprime commented:
Thanks for this PR @matthewbauer ! After 220 derivations & a couple of days  on my small ARM board, I've got a statically compiled armv7l nix! 

[nix-2.2-armv7l.tar.gz](https://github.com/NixOS/nixpkgs/files/2932313/nix-2.2-armv7l.tar.gz)

I did have to edit the bash derivation though:

```diff
diff --git a/pkgs/shells/bash/4.4.nix b/pkgs/shells/bash/4.4.nix
index e5e33c7..a61801a 100644
--- a/pkgs/shells/bash/4.4.nixmaybe this is because it can't create user namespaces onmaybe this is because it can't create user namespaces on this board and it defaults to  this board and it defaults to 
+++ b/pkgs/shells/bash/4.4.nix
@@ -28,7 +28,7 @@ stdenv.mkDerivation rec {
     sha256 = "1jyz6snd63xjn6skk7za6psgidsd53k05cr3lksqybi0q6936syq";
   };
 
-  hardeningDisable = [ "format" ];
+  hardeningDisable = [ "format" ] ++ optional (stdenv.hostPlatform.libc == "musl") "pie";
 
   outputs = [ "out" "dev" "man" "doc" "info" ];
```

now I can run it but nix seems to ignore the `--store` argument (it tries to find `/nix`), maybe it can't setup the chroot and defaults to the normal store, but I have this:
```bash
$ zcat /proc/config.gz | egrep 'USER_NS|SECCOMP'
CONFIG_USER_NS=y
CONFIG_HAVE_ARCH_SECCOMP_FILTER=y
CONFIG_SECCOMP_FILTER=y
CONFIG_SECCOMP=y

$ sudo mkdir -p /nix/store

$ ./nix run --store ~/my-nix nixpkgs.nix nixpkgs.bashInteractive
error: getting status of '/nix/store/3azwf31ib816jq403qrymzw02hrlj5fk-nix-2.2-armv7l-unknown-linux-musleabihf': No such file or directory

$ ./nix show-config | grep '/nix/store'
build-hook = /nix/store/3azwf31ib816jq403qrymzw02hrlj5fk-nix-2.2-armv7l-unknown-linux-musleabihf/libexec/nix/build-remote
sandbox-paths = /bin/sh=/nix/store/l31h4is1irdw1zfld2qf7l0l4w98ylrh-busybox-1.30.1-armv7l-unknown-linux-musleabihf/bin/busybox
```

i'll keep poking around, thought i'd post the built binary here if it helps anyone

matthewbauer commented:
You need to provide a Nix channel / file for it to use. So most likely you need something like this:

```
export NIX_DATA_DIR=$(nix-build '<nixpkgs>' -A nix)/share
./nix run --store ~/my-nix -f channel:nixpkgs-unstable nixpkgs.nix nixpkgs.bashInteractive
```

Both the ```--store``` and ```-f``` flags would be nice to provide defaults ofr.

NIX_DATA_DIR is needed to pick up some "booter" expressions in corepkgs. You can use the native Nix for this to avoid rebuliding.

illegalprime commented:
@matthewbauer that worked! i'm currently building with a static nix in a custom nix-store :tada: !
_but_ in order to do that I did have to include these files in my root's `/nix/store`, even though there's no other nix files present on the system:

```
$ find /nix/store ! -type d
/nix/store/l31h4is1irdw1zfld2qf7l0l4w98ylrh-busybox-1.30.1-armv7l-unknown-linux-musleabihf/bin/ash
/nix/store/l31h4is1irdw1zfld2qf7l0l4w98ylrh-busybox-1.30.1-armv7l-unknown-linux-musleabihf/bin/sh
/nix/store/l31h4is1irdw1zfld2qf7l0l4w98ylrh-busybox-1.30.1-armv7l-unknown-linux-musleabihf/bin/busybox
/nix/store/3azwf31ib816jq403qrymzw02hrlj5fk-nix-2.2-armv7l-unknown-linux-musleabihf/libexec/nix/build-remote

$ ls root/nix/store | wc -l
814
```

the files in my `/nix/store` also have incorrect permissions (they're writable) so it looks like it's just a hard-coded path in the static nix.

so, the static nix is fully portable except that it references those two paths.
maybe a unit test that checks that no buildInputs are runtime dependencies?

matthewbauer committed:
don’t use stackprotector for x86_32

matthewbauer committed:
systems: correct qemu architectures

ppc64le and ppc64 are different targets in the configure script. We
can’t use the same one.

TODO: canonicalize similar ones based on qemu’s configure script.

Ericson2314 commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -295,6 +295,10 @@ stdenv.mkDerivation {
     ''
 
+    + optionalString targetPlatform.isx86_32 ''
+      hardening_unsupported_flags+=" stackprotector"
```
This is temporary I assume?

matthewbauer replied:
Maybe... binutils and musl seemed to get stuck. I don't think anyone supports i686 very well any more.

Ericson2314 replied:
that combination or in isolation? `with targetPlatform.isx86_32 && targetPlatform.isMusl` I'd be more comfortable with that.

matthewbauer commented:
I have a repo that holds some scripts to build static nix here:

https://github.com/matthewbauer/static-nix

It makes it easy to cross build Nix to a few architectures like:

- aarch64
- armv6l
- armv7a
- powerpc64le
- i486
- i686

matthewbauer committed:
systems: fix emulator identity

Ericson2314 commented on lib/systems/default.nix:
```diff
@@ -102,7 +102,7 @@ rec {
         if final.parsed.kernel.name == pkgs.stdenv.hostPlatform.parsed.kernel.name &&
            pkgs.stdenv.hostPlatform.isCompatible final
-        then "${pkgs.runtimeShell} -c"
+        then "${pkgs.runtimeShell} -c '$@' --"
```
You might want `"${pkgs.runtimeShell} -c '\"$@\"' --"` actually, but I'm just guessing.

matthewbauer committed:
fix shell quoting

thanks @ericson2314

nixos-discourse commented:
This pull request has been mentioned on **Nix community**. There might be relevant details there:

https://discourse.nixos.org/t/crosscompilation-to-musl32-problems/3110/3

domenkozar commented:
I've polished a few commits from here and pushed to master to fix builds of bohemgc, brotli, and busybox.

What's left seems to be: curl, boost and openssl.

domenkozar commented:
Fixed also curl, boost, opeensl and libsodium.

Last one standing before nix is nghttp2:

```
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking for grep that handles long lines and -e... (cached) /nix/store/6zcfsp4rkn7pyc04zy0jww3p5lfhvvh4-gnugrep-3.3/bin/grep
checking for egrep... /nix/store/6zcfsp4rkn7pyc04zy0jww3p5lfhvvh4-gnugrep-3.3/bin/grep -E
checking if OS is AIX (to define _ALL_SOURCE)... no
checking if _THREAD_SAFE is already defined... no
checking if _THREAD_SAFE is actually needed... no
checking if _THREAD_SAFE is onwards defined... no
checking if _REENTRANT is already defined... no
/nix/store/gkq6rh8nj8i7prh4c9rawhwrk5g20cfx-x86_64-unknown-linux-musl-binutils-2.31.1/bin/x86_64-unknown-linux-musl-ld: cannot find -lz
collect2: error: ld returned 1 exit status
make[3]: *** [Makefile:1414: nghttp] Error 1
make[3]: Leaving directory '/build/nghttp2-1.39.2/src'
make[2]: *** [Makefile:2678: all-recursive] Error 1
make[2]: Leaving directory '/build/nghttp2-1.39.2/src'
make[1]: *** [Makefile:567: all-recursive] Error 1
make[1]: Leaving directory '/build/nghttp2-1.39.2'
make: *** [Makefile:476: all] Error 2
builder for '/nix/store/ba1w3wbdp4k02pkanhwcpyh2cv2qj2wh-nghttp2-1.39.2-x86_64-unknown-linux-musl.drv' failed with exit code 2
```

Any ideas?

domenkozar commented:
Fixed also zlib and nghttp2.

domenkozar commented:
It gets all the way to nix now:

```
Updating Autotools / GNU config script to a newer upstream version: ./config/config.guess
configuring
chmod: missing operand after 'u+w'
Try 'chmod --help' for more information.
builder for '/nix/store/17hykiklqbk2dyq3jcj6pdb9b32sl9yk-nix-2.3-x86_64-unknown-linux-musl.drv' failed with exit code 1
```

domenkozar commented:
I think we should close this one as it has conflicts. I haven't checked cross compilation (certainly some patches from this PR is missing), but at least Nix compiles statically in https://github.com/NixOS/nixpkgs/pull/70024

fzakaria commented:
Is there a way to get the static binary without flake ?
Can it be published as a release artifact on GitHub ?

domenkozar commented:
It is built by CI, so it just needs that some adds an artifact.

fzakaria commented:
That's so cool. A lot of guides or tutorials could use it instead for a very light installation or demo.

fzakaria commented:
Off topic: @domenkozar why is multi user install becoming the default ?

domenkozar commented:
See https://github.com/NixOS/nix/pull/4181 for discussion

matthewbauer commented:
@fzakaria You can now download static nix via Hydra at:

https://hydra.nixos.org/job/nix/master/buildStatic.x86_64-linux/latest/download/2/nix

We don't have any official release on nixos.org yet though.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/where-can-i-get-a-statically-built-nix/34253/3
